// @unocss/preset-mini@0.45.30/dist/variants.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const _default = require('./shared/preset-mini.5c501ea4.cjs');
require('./shared/preset-mini.2d0396b3.cjs');
require('@unocss/core');
require('./shared/preset-mini.6d3bd985.cjs');



exports.calcMaxWidthBySize = _default.calcMaxWidthBySize;
exports.partClasses = _default.partClasses;
exports.variantBreakpoints = _default.variantBreakpoints;
exports.variantColorsMediaOrClass = _default.variantColorsMediaOrClass;
exports.variantCombinators = _default.variantCombinators;
exports.variantCssLayer = _default.variantCssLayer;
exports.variantCustomMedia = _default.variantCustomMedia;
exports.variantImportant = _default.variantImportant;
exports.variantInternalLayer = _default.variantInternalLayer;
exports.variantLanguageDirections = _default.variantLanguageDirections;
exports.variantNegative = _default.variantNegative;
exports.variantPrint = _default.variantPrint;
exports.variantPseudoClassFunctions = _default.variantPseudoClassFunctions;
exports.variantPseudoClassesAndElements = _default.variantPseudoClassesAndElements;
exports.variantScope = _default.variantScope;
exports.variantSelector = _default.variantSelector;
exports.variantSupports = _default.variantSupports;
exports.variantTaggedPseudoClasses = _default.variantTaggedPseudoClasses;
exports.variantVariables = _default.variantVariables;
exports.variants = _default.variants;
